File: urbanairship-core/src/test/java/com/urbanairship/locale/LocaleManagerTest.java
package com.urbanairship.locale;

import android.content.Context;

import com.urbanairship.BaseTestCase;
import com.urbanairship.TestApplication;

import org.junit.Before;
import org.junit.Test;

import java.util.Locale;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

/**
 * Tests for {@link LocaleManager}
 */
public class LocaleManagerTest extends BaseTestCase {

    private LocaleManager localeManager;
    private Context context;

    @Before
    public void setup() {
        context = TestApplication.getApplication();
        localeManager = new LocaleManager(context);
    }

    @Test
    public void testGetLocale() {
        Locale de = new Locale("de");
        context.getResources().getConfiguration().setLocale(de);

        assertEquals(de, localeManager.getDefaultLocale());
    }

    @Test
    public void testNotifyLocaleChanged() {
        Locale en = new Locale("en");
        context.getResources().getConfiguration().setLocale(en);

        LocaleChangedListener listener = mock(LocaleChangedListener.class);
        localeManager.addListener(listener);

        localeManager.notifyLocaleChanged();

        verify(listener).onLocaleChanged(en);
    }

}

File: urbanairship-core/src/test/java/com/urbanairship/locale/LocaleManagerTest.kt
package com.urbanairship.locale

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.urbanairship.PreferenceDataStore
import com.urbanairship.util.LocaleCompat
import java.util.Locale
import app.cash.turbine.test
import io.mockk.mockk
import io.mockk.verify
import kotlinx.coroutines.test.TestResult
import kotlinx.coroutines.test.runTest
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Tests for [LocaleManager]
 */
@RunWith(AndroidJUnit4::class)
public class LocaleManagerTest {

    private var context: Context = ApplicationProvider.getApplicationContext()
    private var localeManager = LocaleManager(context, PreferenceDataStore.inMemoryStore(context))

    @Test
    public fun testGetLocale() {
        val de = LocaleCompat.of("de")
        context.resources.configuration.setLocale(de)

        Assert.assertEquals(de, localeManager.locale)
    }

    @Test
    public fun testLocaleUpdatesFlow(): TestResult = runTest {
        localeManager.localeUpdates.test {
            val en = LocaleCompat.of("en")
            context.resources.configuration.setLocale(en)

            localeManager.onDeviceLocaleChanged()

            Assert.assertEquals(en, awaitItem())
        }
    }
}
